segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1435 @@
+        }
+    }
+
+    /**
+     * Seals a StreamSegment. No further modifications are allowed on the StreamSegment after this operation completes.
+     *
+     * @param handle  A read-write SegmentHandle that points to a Segment to Seal.
+     * @param timeout Timeout for the operation.
+     * @return A CompletableFuture that, when completed, will indicate that the operation completed. If the operation
+     * failed, it will contain the cause of the failure. Notable exceptions:
+     * <ul>
+     * <li> StreamSegmentNotExistsException: When the given Segment does not exist in Storage.
+     * <li> StorageNotPrimaryException: When this Storage instance is no longer primary for this Segment (it was fenced out).
+     * </ul>
+     * @throws IllegalArgumentException If handle is read-only.
+     */
+    @Override
+    public CompletableFuture<Void> seal(SegmentHandle handle, Duration timeout) {
+        checkInitialized();
+        return execute(() -> {
+            long traceId = LoggerHelpers.traceEnter(log, "seal", handle);
+            Preconditions.checkNotNull(handle, "handle");
+            String streamSegmentName = handle.getSegmentName();
+            Preconditions.checkNotNull(streamSegmentName, "streamSegmentName");
+            Preconditions.checkArgument(!handle.isReadOnly(), "handle");
+
+            try (MetadataTransaction txn = metadataStore.beginTransaction()) {
+                SegmentMetadata segmentMetadata = (SegmentMetadata) txn.get(streamSegmentName);
+                // Validate preconditions.
+                checkSegmentExists(streamSegmentName, segmentMetadata);
+
+                if (segmentMetadata.getOwnerEpoch() > this.epoch) {
+                    throw new StorageNotPrimaryException(streamSegmentName);
+                }
+
+                // seal if it is not already sealed.
+                if (!segmentMetadata.isSealed()) {
+                    segmentMetadata.setSealed(true);
+                    txn.update(segmentMetadata);
+                    txn.commit();
+                }
+
+                log.debug("{} seal - segment={}.", logPrefix, handle.getSegmentName());
+                LoggerHelpers.traceLeave(log, "seal", traceId, handle);
+                return null;
+            } catch (StorageMetadataWritesFencedOutException ex) {
+                throw new StorageNotPrimaryException(streamSegmentName, ex);
+            }
+        });
+    }
+
+    /**
+     * Concatenates two StreamSegments together. The Source StreamSegment will be appended as one atomic block at the end
+     * of the Target StreamSegment (but only if its length equals the given offset), after which the Source StreamSegment
+     * will cease to exist. Prior to this operation, the Source StreamSegment must be sealed.
+     *
+     * @param targetHandle  A read-write SegmentHandle that points to the Target StreamSegment. After this operation
+     *                      is complete, this is the surviving StreamSegment.
```

Comment:
Surviving is a funny term, what about "After this operation completes, the target segment incorporates the source segment, and only the target segment remains."

Reply:
IDE magic, it copied exact comment from already present Storage interface.

Reply:
https://github.com/pravega/pravega/issues/4886